rfcs/0092-plan-dynamism.md

```diff
@@ -0,0 +1,258 @@
+END
+```
+
+The way you would use a derivation that outputs a derivation to `out` is then
+as such:
+```nix
+{ pkgs, drv }:
+
+pkgs.runCommand "example" {} ''
+  ls ${builtins.outputOf drv.out "out"} > $out
+''
+```
+
+Given a path to a derivation that might not yet be built, `builtins.outputOf`
+gives us the path to an output of it.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- We add a bit of complexity to Nix.
+- There is currently no way of getting a "derivation object"
+  as you do from `builtins.derivation` with `builtins.outputOf`.
+  There are reasons for why this can't be done, mainly that you
+  don't actually know the attributes the built derivation will have,
+  but it might be an ergonomic issue.
+- This is for many things not an alternative to IFD, since we
+  still can not build derivations and then use them at evaluation
+  time, meaning that you can't have an attribute set whose contents
+  are determined by some build, and then access that attribute set
+  outside of build that dependens on that derivation.
+- We unfortunately expose the `text` `outputHashMode` to users.
+  Preferably this should be removed entirely, in addition to `flat`,
+  and everything should just use `recursive`.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Restrict ourselves to a subset of what we can do with this RFC,
+  and implement that using only Recursive Nix without making use
+  of this RFC.
+  Notably, we can still run `nix build` at the end of builds. This isn't as great,
+  since 1) the daemon will consider the build doing `nix build`
+  as an active build, 2) it messes with logging, often the log
+  of a failing inner build will not be easily accessible, and
+  3) we can't actually have derivations that output derivations.
+- Do nothing, and continue to have no good answer for large builds like Linux and Chromium.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- The exact way the outputs refer to the replacement derivations / their outputs is subject to bikeshedding.
+- Do we need the new CLI?
+- Can we make `builtins.outputOf` more ergonomic?
```

Comment:
This is a good question, since whether a package uses plan dynamism should be an implementation detail of that package that doesn't get exposed to the consumers of that package. So a derivation that uses `hello` shouldn't have to do `${builtins.outputOf hello.out "out"}/bin/hello` if the hello package uses plan dynamism. Instead `${hello}/bin/hello` should work as expected.

Reply:
This should be the responsibility of the package expressions that use plan dynamism. Ie a package output must not be a derivation path, whereas derivations in general can do anything. We'll have to stop conflating derivations and packages.
I would define a package as an attrset with attributes like `meta`, `outPath` and possibly `bin`, etc, whereas a derivation is a more low level concept represented by attrsets that may have little more than outputs in them.

[This](https://github.com/NixOS/rfcs/pull/92#discussion_r671741003) is a taste of what a package using `outputOf` might look like, although I think we'll build abstractions on top of `mkPackage` to facilitate specific use cases like the specific language integrations that use plan dynamism.